dali/pipeline/operators/reader/reader_op.h

```diff
@@ -197,60 +197,28 @@ class DataReader : public Operator<Backend> {
 
   // GPUBackend operators
   void Run(DeviceWorkspace* ws) override {
-    {
-      std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
-      StartPrefetchThread();
-
-      if (batch_stop_) batch_stop_ = false;
-    }
-
-    {
-      // block all other worker threads from taking the prefetch-controller lock
-      std::unique_lock<std::mutex> worker_lock(worker_mutex_);
-
-      if (!prefetch_ready_workers_) {
-        // grab the actual prefetching lock
-        std::unique_lock<std::mutex> prefetch_lock(prefetch_access_mutex_);
-
-        // Wait until prefetch is ready
-        consumer_.wait(prefetch_lock, [&]() { return prefetched_batch_ready_; });
+    StartPrefetchThread();
 
-        if (prefetch_error_) {
-          DALI_FAIL("Prefetching failed: " + dali::string(*prefetch_error_));
-        }
-
-        // signal the other workers we're ready
-        prefetch_ready_workers_ = true;
+    // grab the actual prefetching lock
+    std::unique_lock<std::mutex> prefetch_lock(prefetch_access_mutex_);
 
-        producer_.notify_one();
-      }
-    }
+    // Wait until prefetch is ready
+    consumer_.wait(prefetch_lock, [&]() { return prefetched_batch_ready_; });
 
-    for (samples_processed_ = 0;
-         samples_processed_.load() < Operator<Backend>::batch_size_;
-         ++samples_processed_) {
-      // consume batch
-      Operator<Backend>::Run(ws);
-      loader_->ReturnTensor(prefetched_batch_[samples_processed_]);
+    if (prefetch_error_) {
+      DALI_FAIL("Prefetching failed: " + dali::string(*prefetch_error_));
     }
 
-    // lock, check if batch is finished, notify
-    {
-      std::unique_lock<std::mutex> lock(prefetch_access_mutex_);
+    producer_.notify_one();
```

Comment:
I'd like to know a little bit of the motivation of the refactoring, and whether this code is unit tested

Reply:
The thread safety needed by the CPU version of the reader is not required by the GPU one
Tested in a python pipeline but not unit tested in C++